### PR TITLE
Couple two-phase flow, evaporation and heat transfer

### PR DESCRIPTION
**Pull request by mschreter**
This PR introduces a coupling term into the heat equation, considering the heat flux due to evaporation:

image

In addition, the evaporative mass flux entering the continuity equation and the level set equation is computed as follows:

image

Therein, the evaporation coefficient alpha_v is computed according to [Hardt & Wondra 2008](https://doi.org/10.1016/j.jcp.2008.02.020).

Two new benchmarks according to Hardt and Wondra 2008 are introduced:

- Film boiling 
animation

image
- Stefan's problem 2

The new file `physical_constants.hpp` collects physical constants such as the Stefan Boltzmann constant needed for the heat transfer equation.

### CODE REVIEW

File: include/meltpooldg/interface/parameters.hpp [hunk 1]
```diff
@@ -384,12 +389,12 @@ namespace MeltPoolDG
...
-                              ExcMessage(
-                                "For the consideration of phase change, the density "
-                                "has to be interpolated consistently with the continuity equation "
-                                "including phase change."));
+                  // AssertThrow(flow.variable_properties_over_interface ==
+                  //"consistent_with_evaporation",
+                  // ExcMessage(
+                  //"For the consideration of phase change, the density "
+                  //"has to be interpolated consistently with the continuity equation "
+                  //"including phase change."));
```
Comment (mschreter) [inline include/meltpooldg/interface/parameters.hpp:397]: I'll consider to reactivate this assert later.

File: include/meltpooldg/level_set/level_set_operator_with_phase_change.hpp [hunk 2]
```diff
@@ -112,50 +110,40 @@ namespace MeltPoolDG::LevelSet
           FECellIntegrator<dim, dim, number> flow_vel(matrix_free,
                                                       flow_vel_dof_idx,
                                                       this->quad_idx);
+
           FECellIntegrator<dim, dim, number> evapor_vel(matrix_free,
                                                         evapor_vel_dof_idx,
                                                         this->quad_idx);
-          if constexpr (dim > 1)
+          for (unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
```
Comment (mschreter) [inline include/meltpooldg/level_set/level_set_operator_with_phase_change.hpp:117]: Nothing relevant changes (I simply removed `if constexpr (dim>1)`)

File: include/meltpooldg/level_set/level_set_operator_with_phase_change.hpp [hunk 3]
```diff
@@ -177,56 +165,48 @@ namespace MeltPoolDG::LevelSet
...
+
+          FECellIntegrator<dim, dim, number> flow_vel(matrix_free,
                                                       flow_vel_dof_idx,
                                                       this->quad_idx);
 
           FECellIntegrator<dim, dim, number> evapor_vel(matrix_free,
                                                         evapor_vel_dof_idx,
                                                         this->quad_idx);
-          if constexpr (dim > 1)
+          for (unsigned int cell = macro_cells.first; cell < macro_cells.second; ++cell)
```
Comment (mschreter) [inline include/meltpooldg/level_set/level_set_operator_with_phase_change.hpp:176]: same here.

File: include/meltpooldg/simulations/film_boiling/film_boiling.hpp [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+      this->attach_periodic_boundary_condition(left_bc, right_bc, 0);
+
+      this->attach_dirichlet_boundary_condition(lower_bc,
+                                                std::make_shared<Functions::ConstantFunction<dim>>(
+                                                  this->parameters.evapor.boiling_temperature + 5.),
+                                                "heat_transfer");
+      // @todo: had to comment out the following line to not get a partitioner error --> maybe its
+      //        the corner nodes have both -- dirichlet and PBC?
+      // this->attach_dirichlet_boundary_condition(
+      // lower_bc, std::make_shared<Functions::ConstantFunction<dim>>(-1), "level_set");
```
Comment (mschreter) [inline include/meltpooldg/simulations/film_boiling/film_boiling.hpp:235]: When I considered both -- DC and PBC -- I got an exception related to a non-matching partitioner, as discussed with @peterrum.

File: include/meltpooldg/evaporation/evaporation_operation.hpp [hunk 5]
```diff
@@ -175,19 +345,36 @@ namespace MeltPoolDG::Evaporation
...
+              VectorizedArray<double> rho_d_rho_d_phi = 1.0;
+
+              if (interpolation_type_parameters == "true")
+                {
+                  rho_d_rho_d_phi =
+                    (ls.get_value(q_index) * material.second.density +
+                     (1. -
+                      ls.get_value(q_index) *
+                        material.first
+                          .density)) / //-----------------------------------------------------------------------------------------------------
```
Comment (peterrum) [inline include/meltpooldg/evaporation/evaporation_operation.hpp:361]: ?

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 6]
```diff
@@ -591,14 +625,8 @@ namespace MeltPoolDG::Flow
...
           DoFTools::make_periodicity_constraints(
             dof_handler, id_in, id_out, direction, ls_hanging_node_constraints);
-          DoFTools::make_periodicity_constraints(
-            dof_handler, id_in, id_out, direction, ls_constraints_dirichlet);
-          DoFTools::make_periodicity_constraints(
-            dof_handler, id_in, id_out, direction, reinit_constraints_dirichlet);
           DoFTools::make_periodicity_constraints(
             dof_handler_evapor, id_in, id_out, direction, evapor_hanging_node_constraints);
-          DoFTools::make_periodicity_constraints(
-            dof_handler, id_in, id_out, direction, temp_constraints_dirichlet);
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:601]: I guess this is related to #148!?

File: include/meltpooldg/heat/heat_transfer_operator.hpp [hunk 7]
```diff
@@ -565,9 +597,17 @@ namespace MeltPoolDG::Heat
...
+              if (level_set_as_heaviside && evapor_data)
                 {
-                  val += density * capacity * temp_vals.get_gradient(q_index) *
-                         velocity_vals.get_value(q_index);
+                  val += compare_and_apply_mask<SIMDComparison::greater_than_or_equal>(
+                    temp_vals.get_value(q_index),
+                    VectorizedArray<double>(evapor_data->boiling_temperature),
+                    evaporation_heat_transfer_coefficient * ls_vals.get_gradient(q_index).norm() *
+                      (temp_vals.get_value(q_index) - evapor_data->boiling_temperature),
+                    0.0);
```
Comment (peterrum) [inline include/meltpooldg/heat/heat_transfer_operator.hpp:610]: I have a bit of a problem to match the implementation with the formula above:
```
   *                 /        d q_s     \
   *              -  | N_a, ---------   |
   *                 \       dT_b^(n+1) /
   *                                     Ω
 ```

File: include/meltpooldg/level_set/level_set_operator_with_phase_change.hpp [hunk 8]
```diff
@@ -235,8 +215,8 @@ namespace MeltPoolDG::LevelSet
     solve(const double dt, VectorType &advected_field)
     {
       VectorType src, rhs;
-      scratch_data.initialize_dof_vector(rhs, this->dof_idx);
-      scratch_data.initialize_dof_vector(src, this->dof_idx);
```
Comment (peterrum) [inline include/meltpooldg/level_set/level_set_operator_with_phase_change.hpp:239]: :+1:

**Review by peterrum — APPROVED** [review]
Looks fine. I left some comments. 

Could you take at the failing tests. I hope we have not introduced new "not robust tests"^^ After that let's proceed and merge and tackle the open issues in follow up PRs!

File: include/meltpooldg/flow/two_phase_flow_problem.hpp (hunk 6 above)
Comment (mschreter) [reply include/meltpooldg/flow/two_phase_flow_problem.hpp:601]: Yes that's related but did not brought any progress. It is probably not needed to setup duplicates of PBC if they are merged in the subsequent lines anyhow.

**Comment by peterrum** [comment]
```
The following tests FAILED:
	  8 - tests/evaporating_droplet_with_heat.mpirun=4.release (Failed)
	  9 - tests/film_boiling_amr.mpirun=4.release (Failed)
	 32 - tests/stefans_problem2_with_flow_and_heat.mpirun=4.release (Failed)
```

@mschreter We have a problem!

**Comment by mschreter** [comment]
> ```
> The following tests FAILED:
> 	  8 - tests/evaporating_droplet_with_heat.mpirun=4.release (Failed)
> 	  9 - tests/film_boiling_amr.mpirun=4.release (Failed)
> 	 32 - tests/stefans_problem2_with_flow_and_heat.mpirun=4.release (Failed)
> ```
> 
> @mschreter We have a problem!

I know ... grrr

File: include/meltpooldg/simulations/stefans_problem/stefans_problem_with_flow_and_heat.hpp [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+#include <meltpooldg/interface/simulationbase.hpp>
+#include <meltpooldg/utilities/utilityfunctions.hpp>
+
+/**
+ * This example is derived from
+ *
+ * Hardt, S., and F. Wondra. "Evaporation model for interfacial flows based on a continuum-field
+ * representation of the source terms." Journal of Computational Physics 227.11 (2008): 5871-5895.
+ *
+ * and represents a simplified example of the denoted "Stefan's Problem 2".
```
[the remark this answers is not in the document]
Comment (mschreter) [reply include/meltpooldg/simulations/stefans_problem/stefans_problem_with_flow_and_heat.hpp]: Done.

File: include/meltpooldg/heat/heat_transfer_operator.hpp (hunk 7 above)
Comment (mschreter) [reply include/meltpooldg/heat/heat_transfer_operator.hpp]: Thanks Peter, this comment was important. I've clarified the implementation and also added the relevant term in the consistent tangent operator.

File: include/meltpooldg/evaporation/evaporation_operation.hpp (hunk 5 above)
Comment (mschreter) [reply include/meltpooldg/evaporation/evaporation_operation.hpp]: Formatting issue ^^. I tried to write readable code ;-).

File: include/meltpooldg/heat/heat_transfer_operator.hpp [hunk 10]
```diff
@@ -565,9 +609,24 @@ namespace MeltPoolDG::Heat
...
+                       velocity_vals.get_value(q_index);
+
+              if (level_set_as_heaviside && evapor_data)
                 {
-                  val += density * capacity * temp_vals.get_gradient(q_index) *
-                         velocity_vals.get_value(q_index);
+                  /*
+                   * compute the heat sink due to evaporation
+                   *          .
+                   *    q_s = m h_v = α_v · h_v · <(T - T_v)> · δ
```
Comment (peterrum) [inline include/meltpooldg/heat/heat_transfer_operator.hpp:620]: I am still a bit puzzled: the derivation of this formula, i.e.,
``` 
  d q_s     
---------   
dT_b^(n+1) 
```
is? Why not simply `q_s = m h_v = α_v · h_v · δ`. Maybe this is dumb question and the answer is obvious!?

**Comment by mschreter** [comment]
> > ```
> > The following tests FAILED:
> > 	  8 - tests/evaporating_droplet_with_heat.mpirun=4.release (Failed)
> > 	  9 - tests/film_boiling_amr.mpirun=4.release (Failed)
> > 	 32 - tests/stefans_problem2_with_flow_and_heat.mpirun=4.release (Failed)
> > ```
> > 
> > 
> > @mschreter We have a problem!
> 
> I know ... grrr

Unfortunately I don't get the tests to run (at least 2 of them) for now. To further proceed,  I would prefer to move the 2 failing ones to `tests_not_robust`. We have the same issues here as we've observed when we were fixing the test suite.

**Review by peterrum — APPROVED** [review]
> Unfortunately I don't get the tests to run (at least 2 of them) for now. To further proceed, I would prefer to move the 2 failing ones to tests_not_robust. We have the same issues here as we've observed when we were fixing the test suite.

Fine with me.

**Comment by mschreter** [comment]
Dear @peterrum, thanks a lot for taking the time to review this PR. From my side this PR is ready, although the failing tests are suboptimal. I am optimistic that we will be able to fix them soon, since the issues are again related to the norms of normal vector / curvature operation.

**Pull request merged**